File: Daily/sbgo/Leetcode752-打开转盘锁.go
package sbgo

func openLock(deadends []string, target string) int {
	start := "0000"
	if start == target {
		return 0
	}
	lock := map[string]bool{}
	for _, s := range deadends {
		lock[s] = true
	}
	if _, ok := lock[start]; ok {
		return -1
	}
	dist := map[string]int{}
	dist[start] = 0
	que := make([]string, 0)
	que = append(que, start)
	for len(que) > 0 {
		t := que[0]
		que = que[1:]
		for i := 0; i < 4; i++ {
			for j := -1; j < 2; j += 2 {
				st := []byte(t)
				st[i] = byte((int(st[i])-int('0')+j+10)%10) + byte('0')
				if _, ok := dist[string(st)]; !ok {
					if _, ok := lock[string(st)]; !ok {
						dist[string(st)] = dist[t] + 1
						if string(st) == target {
							return dist[string(st)]
						}
						que = append(que, string(st))
					}
				}
			}
		}
	}
	return -1
}

File: Daily/PY/Leetcode752-打开转盘锁.py
from collections import defaultdict
from typing import List
class Solution:
    def openLock(self, deadends: List[str], target: str) -> int:
        start = '0000'
        if start == target:return 0
        lock = set()
        for s in deadends: lock.add(s)
        if start in lock: return -1
        que = [start]
        dist = defaultdict(int)
        dist[start] = 0
        while que:
            t = que.pop(0)
            for i in range(4):
                for j in range(-1,2,2):
                    st = [x for x in t]
                    st[i] = str((ord(st[i]) - ord('0') + j + 10) % 10)
                    st = ''.join(st)
                    if st not in dist and st not in lock:
                        dist[st] = dist[t] + 1
                        if st == target:
                            return dist[st]
                        que.append(st)
        return -1